grails-app/domain/org/pih/warehouse/order/Order.groovy

```diff
@@ -31,13 +36,13 @@ class Order implements Serializable {
 
     def publishRefreshEvent = {
         if (isPurchaseOrder) {
-            publishEvent(new RefreshOrderSummaryEvent(this))
+            applicationEventPublisher?.publishEvent(new RefreshOrderSummaryEvent(this))
```

Comment:
Marking this ticket for https://github.com/openboxes/openboxes/labels/tech-huddle.

Reply:
I want to make sure this is the right approach and then backport this to develop.

Reply:
To clarify what I mean, I think the autowired approach is fine and probably the right approach here but I just wanted to point out that the reason "publishEvent" works on its own is not standard injection but rather groovy magic.

For every controller/domain/service class, the spring events plugins adds the publishEvent as a closure.
https://github.com/robfletcher/grails-spring-events/blob/master/SpringEventsGrailsPlugin.groovy#L107

And the publishEvent closure invokes the publishEvent method on the asynEventPublisher.  
https://github.com/robfletcher/grails-spring-events/blob/master/SpringEventsGrailsPlugin.groovy#L113

We could do something similar here just to make it work out of the box and not require a ton of new service injections. We'd just need to figure out where that magic goes in a Grails 3 application (Bootstrap.groovy). 

I don't know exactly why this magic is no longer happening in Grails 3. Are we no longer using the spring events plugin? Are we using the standard Spring Events API? Anyway, I'm open to suggestions on the right approach. 

And for the record, in Grails 3 (post-migration) and beyond we will be looking into other technologies for asynchronous programming (https://async.grails.org/latest/guide/index.html) and event publishing (https://async.grails.org/latest/guide/index.html#events), namely 

* Reactor 
* RxJava 
* GPars  

or a message queue approach using 

* RabbitMQ https://guides.grails.org/grails-rabbitmq/guide/index.html
* Redis https://github.com/grails/grails-redis
* Kafka https://guides.grails.org/grails-micronaut-kafka/guide/index.html

You can read more about that here  https://docs.grails.org/latest/guide/async.html

Exciting times lie ahead.

Reply:
from what I've read - in Grails 1.3.9 `publishEvent` was a member of `GrailsApplication` object and you could refer to it in any controller/service/domain and in Grails 3 it got updated to use `ApplicationEventPublisher` which is a spring bean from spring context, so the only way I see it working is to inject it into the domain using `@Autowired`.

Had some part of Redis at the university and also saw some usage of Reactor, so I'm also excited if we ever go this way :)